### ISSUE TITLE
cannot infer type for `E`

### ISSUE THREAD

**Issue by grissiom** [Geal/nom#503]
Code:
``` rust
#[macro_use]
extern crate nom;
#[macro_use]
extern crate lazy_static;
extern crate regex;

struct Node<'a> {
    prefix: Option<&'a str>,
    content: &'a str,
}

fn is_char(ch: char) -> bool {
    true
}

named!(parse_node<&str, Node>,
       do_parse!(
           prefix: opt!(re_find_static!(r"(\s*|(?m://.*$))*")) >>
           content: take_while_s!(is_char) >>
           (Node {prefix, content})
       )
      );

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        println!(parse_node("   // asdf  \n content"));
    }
}
```

`cargo build` give the error:
```
% cargo build
   Compiling parse-toy v0.1.0 (file:///home/grissiom/project/rust-projs/parse-toy)
error[E0282]: type annotations needed
  --> src/lib.rs:16:1
   |
16 | / named!(parse_node<&str, Node>,
17 | |        do_parse!(
18 | |            prefix: opt!(re_find_static!(r"(\s*|(?m://.*$))*")) >>
19 | |            content: take_while_s!(is_char) >>
20 | |            (Node {prefix, content})
21 | |        )
22 | |       );
   | |________^ cannot infer type for `E`
   |
   = note: this error originates in a macro outside of the current crate

error: aborting due to previous error

error: Could not compile `parse-toy`.

To learn more, run the command again with --verbose.
```

But if I remove the `Option` of the `Node.prefix` and `opt!`, the problem gone. It seems that this is related to #498.

**Comment by Geal** [comment]
this should be fixed in #484, please test again with nom master

**Comment by grissiom** [comment]
Nope, it still doesn't work:

```
    Updating git repository `https://github.com/Geal/nom.git`
   Compiling nom v2.2.1 (https://github.com/Geal/nom.git#d759b3ca)
   Compiling parse-toy v0.1.0 (file:///home/grissiom/project/rust-projs/parse-toy)
error[E0282]: type annotations needed
  --> src/lib.rs:16:1
   |
16 | / named!(parse_node<&str, Node>,
17 | |        do_parse!(
18 | |            prefix: opt!(re_find_static!(r"(\s*|(?m://.*$))*")) >>
19 | |            content: take_while_s!(is_char) >>
20 | |            (Node {prefix, content})
21 | |        )
22 | |       );
   | |________^ cannot infer type for `E`
   |
   = note: this error originates in a macro outside of the current crate

error: aborting due to previous error

error: Could not compile `parse-toy`.

To learn more, run the command again with --verbose.
```

**Comment by Geal** [comment]
alright, I found where it comes from, that one is easy to fix

**Comment by grissiom** [comment]
It is fixed by commit 6c15cc24f1653. So I'll close it. Thanks!
